src/include/migraphx/op/group_query_attention.hpp

```diff
@@ -0,0 +1,572 @@
+                              const std::size_t* pos_ids,
+                              gqa_parameters parameters) const
+    {
+        const std::size_t batch_size             = parameters.batch_size;
+        const std::size_t sequence_length        = parameters.sequence_length;
+        const std::size_t n_heads                = parameters.num_heads;
+        const std::size_t head_size              = parameters.head_size;
+        const std::size_t head_stride            = parameters.head_stride;
+        const std::size_t seq_stride             = parameters.seq_stride;
+        const std::size_t batch_stride           = parameters.batch_stride;
+        const std::size_t position_ids_use_batch = parameters.position_ids_use_batch;
+        const std::size_t rotary_emb_dim         = parameters.rotary_embedding_dim;
+        const std::size_t half_rotary_emb_dim    = rotary_emb_dim / 2;
+
+        const std::size_t loop_len = batch_size * sequence_length * n_heads;
+        par_for(loop_len, [&](const auto idx) {
+            const std::size_t b            = (idx / n_heads) / sequence_length;
+            const std::size_t s            = (idx / n_heads) % sequence_length;
+            const std::size_t n            = idx % n_heads;
+            const std::size_t block_offset = b * batch_stride + s * seq_stride + n * head_stride;
+            auto input_data                = input + block_offset;
+            auto output_data               = output + block_offset;
+
+            // Cache is (M, H/2) or (M, rotary_embedding_dim/2)
+            const std::size_t position_id =
+                position_ids_use_batch ? pos_ids[b * sequence_length + s] : pos_ids[0] + s;
+            const std::size_t cache_offset = position_id * half_rotary_emb_dim;
+            auto cos_data                  = cos_cache + cache_offset;
+            auto sin_data                  = sin_cache + cache_offset;
+
+            std::size_t cache_idx = 0;
+            float sign            = 0.0;
+            std::size_t j         = 0;
+            for(std::size_t i = 0; i < rotary_emb_dim; i++)
+            {
+                if(interleaved)
+                {
+                    cache_idx = (i / 2) % half_rotary_emb_dim;
+                    sign      = (i % 2 == 0) ? -1.0 : 1.0;
+                    j         = (i % 2 == 0) ? i + 1 : i - 1; // i - sign
+                }
+                else
+                {
+                    cache_idx = i % half_rotary_emb_dim;
+                    sign      = (i < half_rotary_emb_dim) ? -1.0 : 1.0;
+                    j         = (i + half_rotary_emb_dim) % rotary_emb_dim;
+                }
+                output_data[i] = input_data[i] * cos_data[cache_idx] +
+                                 sign * input_data[j] * sin_data[cache_idx];
+            }
+            for(std::size_t i = rotary_emb_dim; i < head_size; i++)
+            {
+                output_data[i] = input_data[i];
+            }
```

Comment:
This can be `std::copy(input_data + rotary_emb_dim, input_data + head_size, output_data + rotary_emb_dim)`.